File: scripts/UpdateTituloAt/updateTituloAt.py
#!/usr/bin/env python3
"""Altera dados de um título (UpdateTituloAT) no webservice Acompanhamento de Títulos da ONR."""

from __future__ import annotations

import json
import os
import sys
from decimal import Decimal
from pathlib import Path

from dotenv import load_dotenv

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from lib.onr_acompanhamento import (  # noqa: E402
    hash_error_hint,
    load_acompanhamento_soap_config,
    load_login_config,
    load_serventia_chave,
    resolve_auth_hash,
)
from lib.onr_env import env_int, env_str, require_env_int  # noqa: E402
from lib.onr_json import to_json_safe  # noqa: E402
from lib.onr_soap import call_operation_from_cfg  # noqa: E402
from lib.onr_update_titulo_at import build_update_titulo_request  # noqa: E402
from lib.onr_zeep_serialize import serialize_result  # noqa: E402

_PREFIX = "ACOMPANHAMENTO_TITULOS_UPDATE_"


def _env_decimal(key: str, default: str = "0") -> Decimal:
    raw = env_str(key, default)
    return Decimal(raw or default)


def _resolve_id_titulo() -> int:
    id_titulo = env_int(f"{_PREFIX}ID_TITULO")
    if id_titulo is None:
        id_titulo = env_int("ACOMPANHAMENTO_TITULOS_ID_TITULO")
    if id_titulo is None:
        raise SystemExit(
            f"Defina {_PREFIX}ID_TITULO ou ACOMPANHAMENTO_TITULOS_ID_TITULO no .env."
        )
    return id_titulo


def load_config() -> dict:
    load_dotenv(ROOT / ".env")

    required = [
        f"{_PREFIX}PROTOCOLO",
        f"{_PREFIX}APRESENTANTE_NOME",
        f"{_PREFIX}DATA_PROTOCOLO",
        f"{_PREFIX}DATA_PREVISAO_ENTREGA",
        f"{_PREFIX}MODO_NOTIFICACAO",
        f"{_PREFIX}INTERESSADO_NOME",
        f"{_PREFIX}NATUREZA_TITULO",
    ]
    missing = [k for k in required if not os.getenv(k)]
    if missing:
        raise SystemExit(f"Variáveis ausentes no .env: {', '.join(missing)}")

    tipo_solicitacao = require_env_int(f"{_PREFIX}TIPO_SOLICITACAO")
    if tipo_solicitacao not in (1, 2):
        raise SystemExit(
            f"{_PREFIX}TIPO_SOLICITACAO deve ser 1 (Prenotação) ou 2 (Exame e Cálculo)."
        )

    modo = (env_str(f"{_PREFIX}MODO_NOTIFICACAO") or "").upper()
    if modo not in ("E", "S"):
        raise SystemExit(f"{_PREFIX}MODO_NOTIFICACAO deve ser E (e-mail) ou S (SMS).")

    apresentante_email = env_str(f"{_PREFIX}APRESENTANTE_EMAIL")
    apresentante_ddd = env_str(f"{_PREFIX}APRESENTANTE_DDD")
    apresentante_telefone = env_str(f"{_PREFIX}APRESENTANTE_TELEFONE")

    if modo == "E" and not apresentante_email:
        raise SystemExit(
            f"Com MODO_NOTIFICACAO=E, defina {_PREFIX}APRESENTANTE_EMAIL no .env."
        )
    if modo == "S" and (not apresentante_ddd or not apresentante_telefone):
        raise SystemExit(
            f"Com MODO_NOTIFICACAO=S, defina {_PREFIX}APRESENTANTE_DDD e "
            f"{_PREFIX}APRESENTANTE_TELEFONE no .env."
        )

    protocolo_raw = env_str(f"{_PREFIX}PROTOCOLO") or ""
    protocolo = "".join(c for c in protocolo_raw if c.isdigit())
    if not protocolo:
        raise SystemExit(
            f"{_PREFIX}PROTOCOLO deve conter apenas dígitos (ex.: 605762). "
            f"Valor atual: {protocolo_raw!r}"
        )
    if len(protocolo) > 11:
        raise SystemExit(
            f"{_PREFIX}PROTOCOLO aceita no máximo 11 dígitos (atual: {len(protocolo)})."
        )

    return {
        "chave": load_serventia_chave(),
        "login_cfg": load_login_config(),
        "id_titulo": _resolve_id_titulo(),
        "protocolo": protocolo,
        "apresentante_nome": env_str(f"{_PREFIX}APRESENTANTE_NOME"),
        "apresentante_email": apresentante_email,
        "apresentante_ddd": apresentante_ddd,
        "apresentante_telefone": apresentante_telefone,
        "apresentante_cpfcnpj": env_str(f"{_PREFIX}APRESENTANTE_CPFCNPJ"),
        "valor_deposito": _env_decimal(f"{_PREFIX}VALOR_DEPOSITO", "0"),
        "valor_emolumentos": _env_decimal(f"{_PREFIX}VALOR_EMOLUMENTOS", "0"),
        "data_protocolo": env_str(f"{_PREFIX}DATA_PROTOCOLO"),
        "data_previsao_entrega": env_str(f"{_PREFIX}DATA_PREVISAO_ENTREGA"),
        "modo_notificacao": modo,
        "interessado_nome": env_str(f"{_PREFIX}INTERESSADO_NOME"),
        "interessado_cpfcnpj": env_str(f"{_PREFIX}INTERESSADO_CPFCNPJ"),
        "natureza_titulo": env_str(f"{_PREFIX}NATUREZA_TITULO"),
        "codigo_verificador": env_str(f"{_PREFIX}CODIGO_VERIFICADOR"),
        "tipo_solicitacao": tipo_solicitacao,
        **load_acompanhamento_soap_config(),
    }


def main() -> int:
    cfg = load_config()
    hash_value = resolve_auth_hash(cfg["chave"], cfg["login_cfg"])
    o_request = build_update_titulo_request(hash_value, cfg)

    print("=== Parâmetros UpdateTituloAT ===")
    print(json.dumps(o_request, ensure_ascii=False, indent=2, default=str))
    print(f"\nEndpoint: {cfg['endpoint']}")

    result = call_operation_from_cfg(cfg, "UpdateTituloAT", o_request)
    response = to_json_safe(serialize_result(result))

    print("\n=== Resposta ===")
    print(json.dumps(response, ensure_ascii=False, indent=2))

    if not response.get("RETORNO"):
        codigo = response.get("CODIGOERRO", "?")
        descricao = response.get("ERRODESCRICAO", "")
        print(f"\nUpdateTituloAT falhou: [{codigo}] {descricao}", file=sys.stderr)
        if codigo == 12:
            print(
                f"Dica: {_PREFIX}ID_TITULO / ACOMPANHAMENTO_TITULOS_ID_TITULO inválido — "
                "confira com ListTitulosAT ou GetTituloAT.",
                file=sys.stderr,
            )
        if codigo in (27, 28):
            print(
                f"Dica: {_PREFIX}PROTOCOLO só com dígitos (até 11), coerente com o título.",
                file=sys.stderr,
            )
        if codigo == 32:
            print(
                "Dica: usuário sem permissão para alterar título (spec cód. 32).",
                file=sys.stderr,
            )
        if codigo == 0 and descricao and "IDMsg" in descricao:
            print(
                "Dica: envie todos os elementos do WSDL (opcionais como string vazia). "
                "Os scripts já fazem isso; se persistir, informe o IDMsg à ONR.",
                file=sys.stderr,
            )
        hint = hash_error_hint(codigo) if isinstance(codigo, int) else None
        if hint:
            print(hint, file=sys.stderr)
        return 1

    print(
        f"\nOK — Título {cfg['id_titulo']} atualizado (protocolo {cfg['protocolo']})."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/UpdateTituloAt/updateTituloAt.js
/**
 * Altera dados de um título (UpdateTituloAT) no webservice Acompanhamento de Títulos da ONR.
 */
import fs from "node:fs";
import soap from "soap";
import { envInt, envStr, requireEnvInt, resolvePath } from "../../lib/onr_env.js";
import {
  hashErrorHint,
  loadAcompanhamentoSoapConfig,
  loadLoginConfig,
  loadServentiaChave,
  resolveAuthHash,
} from "../../lib/onr_acompanhamento.js";
import { buildUpdateTituloRequest } from "../../lib/onr_update_titulo_at.js";

const PREFIX = "ACOMPANHAMENTO_TITULOS_UPDATE_";

function envDecimal(key, defaultValue = "0") {
  const raw = envStr(key, defaultValue);
  return Number(raw ?? defaultValue);
}

function resolveIdTitulo() {
  let idTitulo = envInt(`${PREFIX}ID_TITULO`);
  if (idTitulo === undefined) {
    idTitulo = envInt("ACOMPANHAMENTO_TITULOS_ID_TITULO");
  }
  if (idTitulo === undefined) {
    throw new Error(
      `Defina ${PREFIX}ID_TITULO ou ACOMPANHAMENTO_TITULOS_ID_TITULO no .env.`
    );
  }
  return idTitulo;
}

function loadConfig() {
  const required = [
    `${PREFIX}PROTOCOLO`,
    `${PREFIX}APRESENTANTE_NOME`,
    `${PREFIX}DATA_PROTOCOLO`,
    `${PREFIX}DATA_PREVISAO_ENTREGA`,
    `${PREFIX}MODO_NOTIFICACAO`,
    `${PREFIX}INTERESSADO_NOME`,
    `${PREFIX}NATUREZA_TITULO`,
  ];
  const missing = required.filter((k) => !process.env[k]);
  if (missing.length) {
    throw new Error(`Variáveis ausentes no .env: ${missing.join(", ")}`);
  }

  const tipoSolicitacao = requireEnvInt(`${PREFIX}TIPO_SOLICITACAO`);
  if (tipoSolicitacao !== 1 && tipoSolicitacao !== 2) {
    throw new Error(
      `${PREFIX}TIPO_SOLICITACAO deve ser 1 (Prenotação) ou 2 (Exame e Cálculo).`
    );
  }

  const modo = (envStr(`${PREFIX}MODO_NOTIFICACAO`) || "").toUpperCase();
  if (modo !== "E" && modo !== "S") {
    throw new Error(`${PREFIX}MODO_NOTIFICACAO deve ser E (e-mail) ou S (SMS).`);
  }

  const apresentanteEmail = envStr(`${PREFIX}APRESENTANTE_EMAIL`);
  const apresentanteDdd = envStr(`${PREFIX}APRESENTANTE_DDD`);
  const apresentanteTelefone = envStr(`${PREFIX}APRESENTANTE_TELEFONE`);

  if (modo === "E" && !apresentanteEmail) {
    throw new Error(
      `Com MODO_NOTIFICACAO=E, defina ${PREFIX}APRESENTANTE_EMAIL no .env.`
    );
  }
  if (modo === "S" && (!apresentanteDdd || !apresentanteTelefone)) {
    throw new Error(
      `Com MODO_NOTIFICACAO=S, defina ${PREFIX}APRESENTANTE_DDD e ` +
        `${PREFIX}APRESENTANTE_TELEFONE no .env.`
    );
  }

  const protocoloRaw = envStr(`${PREFIX}PROTOCOLO`) || "";
  const protocolo = protocoloRaw.replace(/\D/g, "");
  if (!protocolo) {
    throw new Error(
      `${PREFIX}PROTOCOLO deve conter apenas dígitos (ex.: 605762). ` +
        `Valor atual: ${JSON.stringify(protocoloRaw)}`
    );
  }
  if (protocolo.length > 11) {
    throw new Error(
      `${PREFIX}PROTOCOLO aceita no máximo 11 dígitos (atual: ${protocolo.length}).`
    );
  }

  return {
    chave: loadServentiaChave(),
    loginCfg: loadLoginConfig(),
    idTitulo: resolveIdTitulo(),
    protocolo,
    apresentanteNome: envStr(`${PREFIX}APRESENTANTE_NOME`),
    apresentanteEmail,
    apresentanteDdd,
    apresentanteTelefone,
    apresentanteCpfcnpj: envStr(`${PREFIX}APRESENTANTE_CPFCNPJ`),
    valorDeposito: envDecimal(`${PREFIX}VALOR_DEPOSITO`, "0"),
    valorEmolumentos: envDecimal(`${PREFIX}VALOR_EMOLUMENTOS`, "0"),
    dataProtocolo: envStr(`${PREFIX}DATA_PROTOCOLO`),
    dataPrevisaoEntrega: envStr(`${PREFIX}DATA_PREVISAO_ENTREGA`),
    modoNotificacao: modo,
    interessadoNome: envStr(`${PREFIX}INTERESSADO_NOME`),
    interessadoCpfcnpj: envStr(`${PREFIX}INTERESSADO_CPFCNPJ`),
    naturezaTitulo: envStr(`${PREFIX}NATUREZA_TITULO`),
    codigoVerificador: envStr(`${PREFIX}CODIGO_VERIFICADOR`),
    tipoSolicitacao,
    ...loadAcompanhamentoSoapConfig(),
  };
}

async function updateTituloAt(cfg, oRequest) {
  const wsdlPath = resolvePath(cfg.wsdlPath);
  if (!fs.existsSync(wsdlPath)) {
    throw new Error(`WSDL não encontrado: ${wsdlPath}`);
  }

  const client = await soap.createClientAsync(wsdlPath, {
    endpoint: cfg.endpoint,
    forceSoap12Headers: false,
    wsdl_options: { timeout: 60000 },
  });

  let response;
  try {
    [response] = await client.UpdateTituloATAsync({ oRequest });
  } catch (err) {
    const status = err?.response?.status;
    const body =
      typeof err?.body === "string"
        ? err.body
        : typeof err?.response?.data === "string"
          ? err.response.data
          : null;

    if (status === 503 || (body && !body.trimStart().startsWith("<"))) {
      throw new Error(
        `Servidor ONR indisponível (HTTP ${status ?? "?"}): ${body ?? err.message}`
      );
    }
    throw err;
  }

  return (
    response?.UpdateTituloATResult ??
    response?.UpdateTituloATResponse?.UpdateTituloATResult ??
    response
  );
}

async function main() {
  const cfg = loadConfig();
  const hash = await resolveAuthHash(cfg.chave, cfg.loginCfg);
  const oRequest = buildUpdateTituloRequest(hash, cfg);

  console.log("=== Parâmetros UpdateTituloAT ===");
  console.log(JSON.stringify(oRequest, null, 2));
  console.log(`\nEndpoint: ${cfg.endpoint}`);

  const response = await updateTituloAt(cfg, oRequest);
  console.log("\n=== Resposta ===");
  console.log(JSON.stringify(response, null, 2));

  if (!response.RETORNO) {
    console.error(
      `\nUpdateTituloAT falhou: [${response.CODIGOERRO}] ${response.ERRODESCRICAO}`
    );
    if (response.CODIGOERRO === 12) {
      console.error(
        `Dica: ${PREFIX}ID_TITULO / ACOMPANHAMENTO_TITULOS_ID_TITULO inválido.`
      );
    }
    if (response.CODIGOERRO === 27 || response.CODIGOERRO === 28) {
      console.error(`Dica: ${PREFIX}PROTOCOLO só com dígitos (até 11).`);
    }
    if (response.CODIGOERRO === 32) {
      console.error("Dica: sem permissão para alterar título (spec cód. 32).");
    }
    if (
      response.CODIGOERRO === 0 &&
      response.ERRODESCRICAO?.includes("IDMsg")
    ) {
      console.error(
        "Dica: envie todos os elementos do WSDL (opcionais como string vazia)."
      );
    }
    const hint = hashErrorHint(response.CODIGOERRO);
    if (hint) console.error(hint);
    process.exit(1);
  }

  console.log(
    `\nOK — Título ${cfg.idTitulo} atualizado (protocolo ${cfg.protocolo}).`
  );
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
